Util/AnnotationLimitProcessor.php

```diff
@@ -44,14 +44,18 @@ public function getRateLimit(Request $request)
         return $best_match;
     }
 
-    public function getRateLimitAlias(Request $request)
+    public function getRateLimitAlias(Request $request, callable $controller)
     {
+        $annotations = $request->attributes->get('_x-rate-limit', array());
```

Comment:
We are repeating this part in each of method. 
I think, that its related to the fact, that current implementation with events implies very dynamic way to work with annotations. From my point of view, that fact push do not use service or need to rebuild service each time as new object. 

Best option as I see - to load all annotation ahead, so we can really use fallback logic with chain of responsibility. Not sure if its possible in symfony, will try to find out. One worry, that it can be breacking change.

Reply:
Instantiating a new object is more expensive in therms of CPU than calling a method (but I think is too soon to consider this kind of optimizations).

Currently symfony does not allow to have dependency-injection containers per url/path, so this kind of `if` checks are inevitable. The same if was done [here](https://github.com/amanto/RateLimitBundle/pull/2/files#diff-9432b63fd61edc833e79324050941460L72) in the previous implementation.

But then based on that if condition was changing the underlying services, by  swapping the rate limit processor with `AnnotationLimitProcessor`... that to me sounds a dirty trick. While the proposed implementation used the well known chain-of-responsibility pattern.